drivers/isl/isl.cpp

```diff
@@ -0,0 +1,350 @@
+int ISL::stop()
+{
+	int result = DevObj::stop();
+
+	if (result != 0) {
+		DF_LOG_ERR("DevObj stop failed");
+		return result;
+	}
+
+	usleep(100000);
+
+	result = devClose();
+
+	if (result != 0) {
+		DF_LOG_ERR("device close failed");
+		return result;
+	}
+
+	usleep(100000);
+	return 0;
+}
+
+//ISL sensor calibration routine:
+//User is required to put sensor at REF_DIST and run this routine to get
+//calibration value.
+//TODO: Make the calculated offset persistent using px4 param system
+int ISL::calibration(void)
+{
+	uint16_t raw_data;
+	float raw_avg = 0;
+	write_reg(0x2f, 0);	//DIST_CAL_OFF_MSB
+	write_reg(0x30, 0);	//DIST_CAL_OFF_LSB
+	usleep(100000);
+	for(uint8_t i = 0; i< 100; i++) {
+		raw_data = read_reg(0xd2);		//DIST_REG_LSB
+		raw_data |= (read_reg(0xd1))<<8; //DIST_REG_MSB
+		raw_avg += float(raw_data)/100;
+	}
+	raw_avg /= 2000;	//convert to meters
+	float offset;
+	offset = (raw_avg - REF_DIST)*2000;
+	if(offset < 0) {
+		return -1;
+	}
+	uint8_t off_lsb, off_msb;
+	off_lsb = (0xFF & uint16_t(offset));
+	off_msb = (0xFF00 & uint16_t(offset))>>8;
+
+	write_reg(0x2f, off_msb);
+	write_reg(0x30, off_lsb);
+	return 0;
+}
+
+//Routine to scan through I2C bus for ISL sensor
+int ISL::_detect()
+{
+	uint8_t dev_id;
+	int result;
+	for(uint8_t i = 0; i <= 3; i++) {
+		result = _setSlaveConfig(_slave_addr,
+			 ISL_BUS_FREQUENCY_IN_KHZ,
+			 ISL_TRANSFER_TIMEOUT_IN_USECS);
+		if (result != 0) {
+			_slave_addr++;
+			continue;
+		}
+		result = DevObj::start();
+		if (result == 0) {
+			dev_id = read_reg(0x00);
+			if(dev_id == 0x0A) {
+				_detected = true;
+				DF_LOG_INFO("Detected ISL sensor @ 0x%x",_slave_addr);
+				return 0;
+			}
+		}
+		_slave_addr++;
+		DevObj::stop();
+	}
+
+	return -1;
+}
+
+// read distance data from ISL sensor
+void ISL::_measure(void)
+{
+	if(!_detected) {
+		return;
+	}
+
+	float distance;
+	//setup for reading sensor
+	write_reg(0xB0, 0x49);
+	uint16_t data_msb, data_lsb, prec;
+	uint16_t offset;
+	uint8_t inter, data_invalid;
+
+	//read data ready flag
+	inter = read_reg(0x69);
+
+	//read sensor offset
+	offset = read_reg(0x30);
+	offset |= (read_reg(0x2f)<<8);
+
+	//read data validity
+	data_invalid = read_reg(0xd0);
+
+	//read corrected data
+	data_lsb = read_reg(0xd2);
+	data_msb = read_reg(0xd1);
+
+	//check if data was ready
+	if(!inter) {
+		DF_LOG_ERR("Bad Distance Data Flags: 0x%x 0x%x", data_invalid, inter);
+		return;
+	}
+	//read quality of data
+	//TODO: send this data to EKF for processing
+	prec = read_reg(0xd4);
+	prec |= ((uint16_t)read_reg(0xd3)) << 8;
+	//read temperature of sensor
+	//TODO: do temperature correction
+	uint8_t temperature;
+	temperature = read_reg(0xe2);
+	//convert distance data to meters
+	distance = ((data_msb<<8) + data_lsb)/2000.0f;
+	//correct distance from errors due to phase-distance inconsistency
+	float phase = (distance / 33.3f ) * M_PI * 2.0f;
+    distance += (1.1637*pow(phase,3) - 3.8654*pow(phase,2) + 1.3796*phase - 0.0436);
```

Comment:
Did you intent to do this math in double precision? Multiply the number 3 and two times instead of using pow